### PR TITLE
Add SSO Models and initial SAML2 views

### PR DESCRIPTION
**Pull request by biyeun**
## Summary
https://dimagi-dev.atlassian.net/browse/SAAS-11606

This is still an inactive part of the system and is still a work in progress. This code can be merged in at any time without affecting production as the changes are 100% isolated from django's auth framework right now.

I will be testing this on `staging` end-to-end with a trial Azure AD account just to verify that the SAML config and certificate handling is properly set up. Security testing will happen at a later stage once the rest of the SSO infrastructure is set up for QA. There's no UI here, just database models and views set up according to `python3-saml`'s docs. I decided to go with the `python3-saml` library over `pysaml2` because of greater flexibility in dealing with certificates and multiple Service Provider <> Identity Provider relationships and documentation that is way easier to follow.

## Feature Flag
None needed. Inactive part of the codebase right now.

## Product Description
No visible changes.

## Safety Assurance
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage
Added some tests for the certificate related tasks and functionality.

### QA Plan
This PR does not need QA as this is an inactive part of the codebase right now. QA will happen once all the UI-related pieces are in place.

### Safety story
Does not interact with django's auth framework right now. 100% isolated, and not even activated without manually adding an `IdentityProvider` in the database.

### Dependency Updates
adds `python3-saml` library

requires some new `brew`/`apt` packages that `xmsec` (a dependency of `python3-saml`) is dependent on:
- `libxml2-dev` (should have already been installed)
- `libxmlsec1-dev`
- `libxmlsec1-openssl`

see https://pypi.org/project/xmlsec/ for full install notes on `apt`/`brew` and other package managers

CommCare Cloud notes: a changelog was already added for the new non-python package requirements and all SaaS environments have these packages installed.

### Rollback instructions
99.9% will not require a rollback, but this does add an initial migration to a new django app, so just be aware of that.
- [ ] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Review by mjriley — COMMENTED** [review]
This looks like a great foundation. I like the code's organization, and I definitely appreciated the atomic commits during this PR. There are a bunch of small issues that I'd like addressed, but looks good otherwise.

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:65]: sure that makes sense!

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:93]: Thanks for that link. That was a good read. I initially approached making sure the SP assertions were also signed because in most of the SAML security articles I came across, that was always mentioned as an area of vulnerability and I wanted to make this setup as secure as possible.

I generated separate signing keys per IdP in the event one gets compromised it doesn't open up other assertions to being sniffed. The TLS point makes some sense though. 

> These requests can also be signed but in practice mostly breaks things when the SP rotates their key. There’s little security benefit since the SAML exchange is all happening over TLS in any modern system.

The plan is to use the rollover cert feature of `python3-saml` that should work around this. Is there any harm in keeping this in place for a super extra layer of security?

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:150]: a quick google search [led me here](https://www.nic.ad.jp/timeline/en/20th/appendix1.html#:~:text=Each%20element%20of%20a%20domain,or%20end%20with%20a%20hyphen.), but then again that's not including the TLD.

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:151]: I always default to using `PROTECT` so that whoever is deleting an object knows that there are other things still dependent on that object so the deletion should be very explicit about making sure everything else is properly considered/addressed.

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:167]: same logic as my comment above. i don't want to make it easy to delete an `IdentityProvider` on purpose

File: corehq/apps/README.rst
Comment (biyeun) [reply corehq/apps/README.rst:141]: yeah I was definitely thinking of adding something like this once I'm sure I got the functionality working 100%. it's a great idea.

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:110]: we are explicit in other places...

File: corehq/apps/sso/models.py
Comment (biyeun) [reply corehq/apps/sso/models.py:152]: yeah, that's actually a cleaner and better idea. Thanks! I'm going to make that change

File: requirements/prod-requirements.in
Comment (biyeun) [reply requirements/prod-requirements.in:13]: this is going to go back into regular requirements, which sucks because people will need to install `libxmlsec1`. I know the author of that article you linked didn't like this library, but there is no clear alternative and I'm not about to write my own library from scratch with zero crypto experience.

File: corehq/apps/sso/utils.py
Comment (biyeun) [reply corehq/apps/sso/utils.py:70]: @mjriley thanks for pointing this out. I will add the `failOnAuthnContextMismatch`. This test is also a great idea. I need to get things working on staging first just to sanity check that the whole architecture is properly set up.

File: corehq/apps/sso/views.py
Comment (biyeun) [reply corehq/apps/sso/views.py:104]: I was going by `python3-saml`'s example django app for a lot of these things, so the terminology came from that. agree the naming could be clearer so I'll update.

File: corehq/apps/sso/views.py
Comment (biyeun) [reply corehq/apps/sso/views.py:32]: great question. this is mostly for testing purposes and will not be the final implementation of this view. I just wanted to get something rudimentary working.

File: corehq/apps/sso/views.py
Comment (biyeun) [reply corehq/apps/sso/views.py:58]: going by `python3-saml`'s [example django app](https://github.com/onelogin/python3-saml/blob/master/demo-django/demo/views.py#L34) I'm guessing that `init_saml_auth(request)` populates the `AuthNRequestID` in `request.session`

File: corehq/apps/sso/models.py
Comment (mjriley) [reply corehq/apps/sso/models.py:150]: Wow, I had no idea. Thanks

File: corehq/apps/sso/models.py
Comment (mjriley) [reply corehq/apps/sso/models.py:151]: Fair, I think I believe PROTECT meant that the child models would be left behind, not that they would prevent deletion of the parent. I think defaulting to PROTECT is fine

File: corehq/apps/sso/views.py
Comment (mjriley) [reply corehq/apps/sso/views.py:58]: Sorry, probably was a bad highlight. I was mainly unclear on why AuthNRequestID needed to be deleted. I don't see the example app doing that, so was this just a concern that that ID might be sensitive and we don't want it leaking to further stages of request processing?

Comment (biyeun) [reply corehq/apps/sso/views.py:58]: the example app does delete it, as that's what I was following:
https://github.com/onelogin/python3-saml/blob/master/demo-django/demo/views.py#L79-L80

File: corehq/apps/sso/utils.py
Comment (biyeun) [reply corehq/apps/sso/utils.py:6]: yep done

File: corehq/apps/sso/tests/base_tests.py
Comment (biyeun) [reply corehq/apps/sso/tests/base_tests.py:8]: done

File: corehq/apps/sso/tests/test_tasks.py
Comment (biyeun) [reply corehq/apps/sso/tests/test_tasks.py:29]: i've added rewriting these tests to my todo

**Comment by biyeun** [comment]
@mjriley ok I think I've responded to most of your comments above with a couple of issues still outstanding. I will postpone addressing this until I've fixed the docker build and gotten this branch to deploy to staging reliably (xmlsec1 issues)

[ ] [Add app-level readme](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r550499252) for SSO
[ ] [Test task behavior at lower level](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r550782726)
[ ] [Add tests to verify that a request with a valid certificate but mismatched entity name is rejected](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r550794912)

Also, adding a note for anyone following that Matt did verify with Jemurai that it is appropriate to have SP certs and the implementation I've set up is what they would expect for a secure setup.

File: corehq/apps/sso/models.py
Comment (millerdev) [inline corehq/apps/sso/models.py:8]: nit: `object` base class is no longer required since Python 3 (it's implied, and all classes are new-style in Python 3)
```suggestion
class IdentityProviderType:
```

This applies to other places throughout this PR.

**Review by millerdev — COMMENTED** [review]
Only one tiny nitpick from me so far.

File: requirements/prod-requirements.in
Comment (biyeun) [inline requirements/prod-requirements.in:2]: @millerdev my plan for separating `sso-requirements` (as it's an optional app) is to allow separate environment-specific requirements files in `<env>/meta.yml`. I've done this for staging here: https://github.com/dimagi/commcare-cloud/pull/4433

If this looks good, I will remove `-r sso-requirements.in` here so that this PR can be safely merged into master (after a final ✅ ) without affecting deploy, a changelog announcement, or requiring a run of `deploy-stack --tags=apt-get` on all environments in order to install `libxmlsec1-dev` and related `apt` requirements for `python3-saml`.

Comment (millerdev) [reply requirements/prod-requirements.in:2]: :+1: Seems like a reasonable plan.

Comment (millerdev) [reply requirements/prod-requirements.in:2]: One question: I assume an error will occur if `ENABLE_SINGLE_SIGN_ON = True` and these requirements are not installed. Would it be possible to make that error message clearly point to resolution steps (special requirements must be installed)?

File: corehq/apps/sso/decorators.py
Comment (biyeun) [reply corehq/apps/sso/decorators.py:10]: added tests

**Comment by biyeun** [comment]
@millerdev and anyone else following the requirements issues on this PR. We are no longer making `sso` an optional app. Instead, I will be making a changelog announcement after this commcare-cloud PR is merged: https://github.com/dimagi/commcare-cloud/pull/4426, and then giving people a week to run the required `deploy-stack` command on their environments. Once a week as passed, I will merge in this PR (assuming it gets a ✅ by that time)

I also want to make sure everyone is clear that this PR is only introducing preliminary test views and `sso` models. It is not introducing any significant changes to active code on hq and is still 100% separated from our auth framework. There is still a ton of work left to get this into a state where it is ready for QA. This work is not intended to become a giant feature branch, as additional changes will all be feature-flagged until QA is complete.

**Comment by biyeun** [comment]
SAML handshake is a go! This is ready to be merged IMHO.

Important notes:
- inactive part of the system / does not touch any existing user workflows
- not integrated with django's auth framework at all so zero security concerns
- views can't even be accessed without an `IdentityProvider` object created in the db. I don't plan to create one on production, just staging, so there's no way someone could accidentally access this on prod.

**Comment by biyeun** [comment]
@mjriley @millerdev re-ping ^_^

**Comment by biyeun** [comment]
please don't be intimidated by the number of commits in PR. the commits are all very small, self-contained, and well-labeled. there were a few commits related to requirements (all well labeled too) that resulted in some back and forth based on a reversal of wanting to make `sso` an optional app. Other than that I expect everything to be straightforward and mostly in direct response to initial feedback. Thanks! I realize this is a big chunk to look at. Hopefully, you will see that nothing significant changed since the initial round of reviews—just re-organization of some code and additional tests.

File: corehq/apps/sso/utils.py
Comment (millerdev) [inline corehq/apps/sso/utils.py:14]: I thought we always require SSL. Why on earth would an authentication provider use plain HTTP without SSL these days?

I can understand why we may want to allow having it turned off in development environments (which I assume would be indicated by `settings.SAML2_DEBUG`). Can we use that setting rather than `request.is_secure()`?

**Review by millerdev — COMMENTED** [review]
I think it would be good for @mjriley to take a pass yet too since he did a fairly detailed review before.

Would like to have an answer to the SSL question before approving.

File: corehq/apps/sso/utils.py
Comment (biyeun) [reply corehq/apps/sso/utils.py:14]: sure! will make that update

Comment (biyeun) [reply corehq/apps/sso/utils.py]: done ✅  @millerdev

**Comment by biyeun** [comment]
yep, thanks @millerdev! @mjriley is currently taking a pass too and will wait on both of our ✅ s before merging

**Review by millerdev — APPROVED** [review]

File: corehq/apps/sso/tests/test_decorators.py
Comment (mjriley) [inline corehq/apps/sso/tests/test_decorators.py:17]: If you want to break precedent, it looks like the 'correct' way to handle this is with [setUpTestData](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.TestCase.setUpTestData). The advantages seem to be:
- don't have to call `super()`, because TestCase doesn't implement `setUpTestData`
- will be called once per class if the database supports transactions, otherwise will occur for every test

No 'cleanup' method exists, because the transaction takes care of the cleanup automatically for you.

It also seems setUpTestData is probably safer in the event that your code throws an exception. [This link](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.SimpleTestCase.databases) claims that that can result in a seg fault on OS X.

File: corehq/apps/sso/views.py
Comment (mjriley) [inline corehq/apps/sso/views.py:145]: does this gracefully handle failure?

File: requirements/requirements.in
Comment (mjriley) [inline requirements/requirements.in:91]: I like that we're using explicit versions here

**Review by mjriley — APPROVED** [review]
Looks good, most of the feedback I left is related to readability and testing. The one thing I think we should think about, however, is whether we should be concerned about sending the raw private keys to the DB.

File: corehq/apps/sso/views.py
Comment (biyeun) [reply corehq/apps/sso/views.py:145]: no, that isn't the goal here as more things might change when I actually integrate this with auth. this is just the initial views

File: corehq/apps/sso/tests/test_decorators.py
Comment (millerdev) [reply corehq/apps/sso/tests/test_decorators.py:20]: Huh, interesting. It's simpler to remember _"Always call super `setUp...` and `tearDown...` methods unless you have an explicit reason not to"_ (it doesn't hurt anything) than _"Always call `super().setUpClass()` but do not call `super().setUp()` (same for `tearDownClass`/`tearDown`)."_

See also https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.SimpleTestCase.databases

File: corehq/apps/sso/models.py
Comment (millerdev) [reply corehq/apps/sso/models.py:95]: > looks like Cloudwatch keeps logs of SQL inserts

Are you sure about that? That would be surprising to me (if only on account of sheer volume, not to mention the security implications).

Comment (biyeun) [reply corehq/apps/sso/models.py:95]: if we are logging raw data from SQL inserts, that seems like a far bigger security risk than storing certs in postgres?

**Comment by biyeun** [comment]
@mjriley thanks for all the helpful feedback. I have added the following to my todo list:
[ ] [handle failures gracefully on SAML views](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r564842196)
[ ] [clarity on certificate abbreviations](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r564648803)
[ ] [clarity on `expiration_in_seconds`](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r564659445)
[ ] [followup on question re: storing certs in postgres](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r564743209)--this will require a broader discussion
[ ] [be more explicit in testing, see](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r564761732)
[ ] [eliminate randomness in tests](https://github.com/dimagi/commcare-hq/pull/28938#discussion_r564769590)

for the sake of not bloating this PR even more, I'll address this separately in a future PR, but leaving this hear so you know what is on the list that I'm keeping track of.

File: corehq/apps/sso/tests/generator.py
Comment (biyeun) [reply corehq/apps/sso/tests/generator.py:15]: followup here https://github.com/dimagi/commcare-hq/pull/29204

**Pull request merged**
